### ISSUE TITLE
Feature: Sorting modules

### ISSUE THREAD

**Issue by mnaoumov** [azat-io/eslint-plugin-perfectionist#359]
### Describe the rule

It should sort members of the modules similarly to the [sort-classes](https://perfectionist.dev/rules/sort-classes) rule

### Code example

See https://stackblitz.com/edit/stackblitz-starters-p1d3rv?file=index.ts which has [recommended-alphabetical](https://perfectionist.dev/configs/recommended-alphabetical) config enabled

I expect

```ts
export function c() {
  a();
}
export function b() {
  a();
}
function a() {}
```

after executing this rule to become

```ts
export function b() {
  a();
}
export function c() {
  a();
}
function a() {}
```

as I think exported functions (like `public` in classes) should be higher than normal functions (like `private` in classes)

### Additional comments

I am surprised I couldn't find any related feature requests. This seems like a must-have rule

### Validations

- [X] Read the [docs](https://perfectionist.dev).
- [X] Check that there isn't [already an issue](https://github.com/azat-io/eslint-plugin-perfectionist/issues) that reports the same bug to avoid creating a duplicate.

**Comment by azat-io** [comment]
Duplicate #53.

It's an interesting idea, but we don't seem to have a way to do it safely.

**Comment by hugop95** [comment]
I have started a prototype of such rule a while back, there are issues that need to be discussed:

## Type of elements sorted

My prototype is called `sort-files`: it doesn't just sort exports, it sorts all elements that you may find in a file: `namespace`, `interface`, `type`, `module`, `class`, `type`, `enum` and `function` (note that for `class`, `enum`, `interface` and `type`, it doesn't sort their content, as we have rules for them, but their declaration position).

## Safety concerns

Safety concerns: as @azat-io says, we need to ensure that compilation succeeds and runtime behavior is left unchanged. We manage this pretty well in `sort-classes` now. We've had a number of issues, but slowly got rid of all the possible cases. 

I think that eventually we can reach that with `sort-files`, but it may take some time and we may encounter unexpected difficulties.
Classes, enums and variables order matter. And you need to consider that you may have to do the sorting at multiple levels, such as within a `namespace`.

## Overlap with other rules

We may think of adding more restrained rules, such as `sort-file-interfaces` (that would sort only interface declarations), or `sort-functions`. However, such rules become incompatible with a bigger `sort-files` rule afterward: this is why `sort-imports` is not handled by `sort-files`, even though it theoretically could (the complexity of `sort-imports` makes it worth putting it in a separate rule however).

I have opened a PR #360 as there has been quite a few issues regarding this in the past, so I think it's at least interesting to see how far we can go.

**Comment by azat-io** [comment]
Thank you for your issue. Added in 950db4e85be61dddeaa983a0ec0f2cb8bd371711.
Released in [v4.0.0](https://github.com/azat-io/eslint-plugin-perfectionist/releases/tag/v4.0.0).

If you'd like, you can support the release with a retweet:
https://x.com/azat_io_en/status/1858853714238271530
